File: react-canvas/src/game/systems/collision.ts
import type { Vector2D } from '../types';
import type { Ship } from '../entities/Ship';
import type { Asteroid } from '../entities/Asteroid';
import type { Bullet } from '../entities/Bullet';

/**
 * Circle-circle collision using squared distance (no sqrt).
 */
export function checkCircleCollision(
  posA: Vector2D,
  radiusA: number,
  posB: Vector2D,
  radiusB: number
): boolean {
  const dx = posA.x - posB.x;
  const dy = posA.y - posB.y;
  const distSq = dx * dx + dy * dy;
  const radSum = radiusA + radiusB;
  return distSq < radSum * radSum;
}

export interface BulletAsteroidHit {
  bulletIndex: number;
  asteroidIndex: number;
}

/**
 * Check all bullet-asteroid collisions.
 * Returns list of hit pairs.
 */
export function checkBulletAsteroidCollisions(
  bullets: Bullet[],
  asteroids: Asteroid[]
): BulletAsteroidHit[] {
  const hits: BulletAsteroidHit[] = [];

  for (let b = 0; b < bullets.length; b++) {
    const bullet = bullets[b];
    if (!bullet.alive) continue;

    for (let a = 0; a < asteroids.length; a++) {
      const asteroid = asteroids[a];
      if (checkCircleCollision(
        bullet.position,
        bullet.radius,
        asteroid.position,
        asteroid.radius
      )) {
        hits.push({ bulletIndex: b, asteroidIndex: a });
        break; // Each bullet can only hit one asteroid
      }
    }
  }

  return hits;
}

/**
 * Check ship-asteroid collisions.
 * Returns index of first asteroid hit, or -1 if none.
 */
export function checkShipAsteroidCollision(
  ship: Ship,
  asteroids: Asteroid[]
): number {
  if (!ship.alive || ship.isInvulnerable) return -1;

  // Use a slightly smaller hitbox for the ship (feels fairer)
  const shipRadius = ship.radius * 0.6;

  for (let a = 0; a < asteroids.length; a++) {
    const asteroid = asteroids[a];
    if (checkCircleCollision(
      ship.position,
      shipRadius,
      asteroid.position,
      asteroid.radius
    )) {
      return a;
    }
  }

  return -1;
}

File: python-pygame/game/systems/collision.py
"""Collision detection system using circle-circle (squared distance)."""

from game.constants import SCREEN_WIDTH, SCREEN_HEIGHT


def circles_collide(x1, y1, r1, x2, y2, r2):
    """Check if two circles collide using squared distance (no sqrt).
    Also handles wrap-around by checking the shortest distance across screen edges."""
    dx = abs(x1 - x2)
    dy = abs(y1 - y2)

    # Account for screen wrapping - use shortest distance
    if dx > SCREEN_WIDTH / 2:
        dx = SCREEN_WIDTH - dx
    if dy > SCREEN_HEIGHT / 2:
        dy = SCREEN_HEIGHT - dy

    dist_sq = dx * dx + dy * dy
    radii_sum = r1 + r2
    return dist_sq < radii_sum * radii_sum


def check_bullet_asteroid_collisions(bullets, asteroids):
    """Check collisions between bullets and asteroids.
    Returns list of (bullet, asteroid) pairs that collided."""
    collisions = []
    for bullet in bullets:
        if not bullet.alive:
            continue
        for asteroid in asteroids:
            if circles_collide(bullet.x, bullet.y, bullet.radius,
                               asteroid.x, asteroid.y, asteroid.radius):
                collisions.append((bullet, asteroid))
                bullet.alive = False
                break  # one bullet hits one asteroid
    return collisions


def check_ship_asteroid_collisions(ship, asteroids):
    """Check if the ship collides with any asteroid.
    Returns the asteroid it collided with, or None."""
    if not ship.alive or ship.invulnerable:
        return None
    for asteroid in asteroids:
        if circles_collide(ship.x, ship.y, ship.radius,
                           asteroid.x, asteroid.y, asteroid.radius):
            return asteroid
    return None
